File: gosdk-develop/crypto/signature.go
package crypto

import (
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"github.com/btcsuite/btcutil/base58"
	"github.com/ethereum/go-ethereum/crypto"
	"github.com/hdevalence/ed25519consensus"
	"github.com/milon-labs/milon-go-sdk/postcard"
	"github.com/pornin/go-fn-dsa/fndsa"
	blst "github.com/supranational/blst/bindings/go"
	"lukechampine.com/blake3"
	"strings"
)

const (
	SignatureSecp256k1Size = 65
	SignatureEd25519Size   = 64
	SignatureBLS12381Size  = 96
	SignatureFnDsa512Size  = 666
)

type SignatureType uint8

const (
	SignatureTypeSecp256k1 SignatureType = iota
	SignatureTypeEd25519
	SignatureTypeBLS12381
	SignatureTypeFnDsa512
)

// NewSignatureFromBytes creates a Signature from raw bytes, auto-detecting the type by length.
func NewSignatureFromBytes(raw []byte) (*Signature, error) {
	switch len(raw) {
	case SignatureEd25519Size:
		return &Signature{
			Variant: SignatureTypeEd25519,
			Bytes:   raw,
		}, nil
	case SignatureBLS12381Size:
		return &Signature{
			Variant: SignatureTypeBLS12381,
			Bytes:   raw,
		}, nil
	case SignatureSecp256k1Size:
		return &Signature{
			Variant: SignatureTypeSecp256k1,
			Bytes:   raw,
		}, nil
	case SignatureFnDsa512Size:
		return &Signature{
			Variant: SignatureTypeFnDsa512,
			Bytes:   raw,
		}, nil
	default:
		return nil, ErrInvalidSignature
	}
}

// NewSignatureFromStringRelaxed parses a hex (with optional 0x prefix) or Base58 string.
func NewSignatureFromStringRelaxed(s string) (*Signature, error) {
	s = strings.TrimSpace(s)

	if sg, err := newSignatureFromHex(s); err == nil {
		return sg, nil
	}

	return newSignatureFromBase58(s)
}
func newSignatureFromHex(s string) (*Signature, error) {
	s = strings.TrimSpace(s)

	hexStr := s
	if len(s) >= 2 && s[0] == '0' && (s[1] == 'x' || s[1] == 'X') {
		hexStr = s[2:]
	}

	buf, err := hex.DecodeString(hexStr)
	if err != nil {
		return nil, fmt.Errorf("invalid hex string: %w", err)
	}

	return NewSignatureFromBytes(buf)
}
func newSignatureFromBase58(b58Str string) (*Signature, error) {
	buf := base58.Decode(b58Str)
	if len(buf) == 0 {
		return nil, errors.New("invalid base58 string")
	}
	return NewSignatureFromBytes(buf)
}

type Signature struct {
	Variant SignatureType
	Bytes   []byte
}

func (sig *Signature) AsBytes() []byte {
	return sig.Bytes
}

func (sig *Signature) ToHex() string {
	return hex.EncodeToString(sig.Bytes)
}

func (sig *Signature) ToBase58() string {
	return base58.Encode(sig.Bytes)
}

// String implements the Stringer interface.
func (sig *Signature) String() string {
	return sig.ToBase58()
}

// ToSecp256k1 converts to a secp256k1 recoverable signature.
func (sig *Signature) ToSecp256k1() ([]byte, error) {
	if sig.Variant != SignatureTypeSecp256k1 {
		return nil, fmt.Errorf("not a secp256k1 signature, actual type: %d", sig.Variant)
	}

	if len(sig.Bytes) != SignatureSecp256k1Size {
		return nil, fmt.Errorf("invalid secp256k1 signature length: %d", len(sig.Bytes))
	}

	return sig.Bytes, nil
}

// ToEd25519 converts to an Ed25519 signature.
func (sig *Signature) ToEd25519() ([]byte, error) {
	if sig.Variant != SignatureTypeEd25519 {
		return nil, fmt.Errorf("not an ed25519 signature, actual type: %d", sig.Variant)
	}

	if len(sig.Bytes) != SignatureEd25519Size {
		return nil, fmt.Errorf("invalid ed25519 signature length: %d", len(sig.Bytes))
	}

	return sig.Bytes, nil
}

// ToBLS12381 converts to a BLS12-381 signature.
func (sig *Signature) ToBLS12381() ([]byte, error) {
	if sig.Variant != SignatureTypeBLS12381 {
		return nil, fmt.Errorf("not a BLS signature, actual type: %d", sig.Variant)
	}

	if len(sig.Bytes) != SignatureBLS12381Size {
		return nil, fmt.Errorf("invalid BLS signature length: %d", len(sig.Bytes))
	}

	return sig.Bytes, nil
}

// ToFnDsa512 converts to an FN-DSA-512 signature.
func (sig *Signature) ToFnDsa512() ([]byte, error) {
	if sig.Variant != SignatureTypeFnDsa512 {
		return nil, fmt.Errorf("not a FN-DSA-512 signature, actual type: %d", sig.Variant)
	}

	if len(sig.Bytes) != SignatureFnDsa512Size {
		return nil, fmt.Errorf("invalid FN-DSA-512 signature length: %d", len(sig.Bytes))
	}

	return sig.Bytes, nil
}

func (sig *Signature) Verify(msg []byte, pubKey *PublicKey) error {
	if uint8(sig.Variant) != uint8(pubKey.Variant) {
		return fmt.Errorf("signature type mismatch: %d vs %d", sig.Variant, pubKey.Variant)
	}

	switch sig.Variant {
	case SignatureTypeSecp256k1:
		return verifySecp256k1(msg, sig.Bytes, pubKey)
	case SignatureTypeEd25519:
		return verifyEd25519(msg, sig.Bytes, pubKey)
	case SignatureTypeBLS12381:
		return verifyBLS12381(msg, sig.Bytes, pubKey)
	case SignatureTypeFnDsa512:
		return verifyFnDsa512(msg, sig.Bytes, pubKey)
	default:
		return ErrInvalidSignature
	}
}

// verifySecp256k1 verifies a secp256k1 signature.
func verifySecp256k1(msg []byte, sigBytes []byte, pubKey *PublicKey) error {
	// 1. Compute message hash
	var msgHash [32]byte
	if len(msg) == 32 {
		copy(msgHash[:], msg)
	} else {
		msgHash = blake3.Sum256(msg)
	}

	// 2. Get public key
	pk, err := pubKey.ToSecp256k1()
	if err != nil {
		return err
	}

	// 3. Verify signature length
	if len(sigBytes) != SignatureSecp256k1Size {
		return fmt.Errorf("invalid secp256k1 signature length: expected %d, got %d", SignatureSecp256k1Size, len(sigBytes))
	}

	// 4. Validate recovery id (V) value range (27/28 for legacy, 0/1 for EIP-155)
	v := sigBytes[len(sigBytes)-1]
	if v != 27 && v != 28 && v != 0 && v != 1 {
		return fmt.Errorf("invalid signature recovery id (V): %d", v)
	}

	// 5. Convert secp256k1 public key to compressed format (33 bytes)
	compressedPubKey := pk.SerializeCompressed()

	// 6. Verify using go-ethereum's verification method
	signatureWithoutV := sigBytes[:len(sigBytes)-1] // take the first 64 bytes (R, S)
	if crypto.VerifySignature(compressedPubKey, msgHash[:], signatureWithoutV) == false {
		return fmt.Errorf("signature verification failed")
	}

	return nil
}

// verifyEd25519 verifies an Ed25519 signature.
func verifyEd25519(msg []byte, sigBytes []byte, pubKey *PublicKey) error {
	pk, err := pubKey.ToEd25519()
	if err != nil {
		return err
	}

	if len(sigBytes) != SignatureEd25519Size {
		return fmt.Errorf("invalid ed25519 signature length: expected %d, got %d", SignatureEd25519Size, len(sigBytes))
	}

	if !ed25519.Verify(pk, msg, sigBytes) {
		return fmt.Errorf("signature verification failed")
	}

	return nil
}

// verifyBLS12381 验证 BLS12-381 签名
func verifyBLS12381(msg []byte, sigBytes []byte, pubKey *PublicKey) error {
	// 1. Get public key
	pk, err := pubKey.ToBLS12381()
	if err != nil {
		return err
	}

	// 2. Verify signature length
	if len(sigBytes) != SignatureBLS12381Size {
		return fmt.Errorf("invalid BLS signature length: expected %d, got %d", SignatureBLS12381Size, len(sigBytes))
	}

	// 3. Recover signature
	sig := new(blst.P2Affine).Uncompress(sigBytes)
	if sig == nil {
		return fmt.Errorf("failed to recover signature")
	}

	// 4. Verify signature - correct parameter order: (checkPub, pk, checkSig, msg, dst)
	if !sig.Verify(true, pk, true, msg, nil) {
		return fmt.Errorf("signature verification failed")
	}

	return nil
}

// verifyFnDsa512 verifies an FN-DSA-512 signature
func verifyFnDsa512(msg []byte, sigBytes []byte, key *PublicKey) error {
	if !fndsa.Verify(key.Bytes, []byte{}, 0, msg, sigBytes) {
		return fmt.Errorf("signature verification failed")
	}

	return nil
}

// VerifyBatch batch-verifies multiple signatures.
// Batch verification is used only when all signatures are Ed25519;
// otherwise each signature is verified individually.
func VerifyBatch(sigs []*Signature, msgs [][]byte, pubKeys []*PublicKey) error {
	if len(sigs) != len(msgs) || len(msgs) != len(pubKeys) {
		return fmt.Errorf("length mismatch: sigs=%d, msgs=%d, pubkeys=%d", len(sigs), len(msgs), len(pubKeys))
	}

	if len(sigs) == 0 {
		return nil
	}

	// Check whether all signatures are Ed25519 type (batch verification requires all Ed25519)
	allEd25519 := true
	for _, sig := range sigs {
		if sig.Variant != SignatureTypeEd25519 {
			allEd25519 = false
			break
		}
	}

	if allEd25519 {
		// Ed25519 supports batch verification
		verifier := ed25519consensus.NewBatchVerifier()

		for i := range sigs {
			edSig, err := sigs[i].ToEd25519()
			if err != nil {
				return fmt.Errorf("failed to convert signature at index %d: %w", i, err)
			}
			edPub, err := pubKeys[i].ToEd25519()
			if err != nil {
				return fmt.Errorf("failed to convert public key at index %d: %w", i, err)
			}

			verifier.Add(edPub, msgs[i], edSig)
		}

		if !verifier.Verify() {
			return fmt.Errorf("batch verification failed: one or more Ed25519 signatures are invalid")
		}

		return nil
	}

	// Mixed or non-Ed25519 types: verify individually
	for i := range sigs {
		if err := sigs[i].Verify(msgs[i], pubKeys[i]); err != nil {
			return err
		}
	}
	return nil
}

func (sig *Signature) MarshalJSON() ([]byte, error) {
	return json.Marshal(sig.ToBase58())
}

func (sig *Signature) UnmarshalJSON(data []byte) error {
	var s string
	if err := json.Unmarshal(data, &s); err != nil {
		return err
	}

	newSig, err := NewSignatureFromStringRelaxed(s)
	if err != nil {
		return fmt.Errorf("invalid signature: %w", err)
	}

	*sig = *newSig
	return nil
}

func (sig *Signature) MarshalPostcard(serializer *postcard.Serializer) error {
	err := serializer.SerializeU32(uint32(sig.Variant))
	if err != nil {
		return fmt.Errorf("failed to serialize signature variant: %w", err)
	}

	serializer.SerializeFixedBytes(sig.Bytes)
	return nil
}

func (sig *Signature) UnmarshalPostcard(deserializer *postcard.Deserializer) error {
	variant, err := deserializer.DeserializeU32()
	if err != nil {
		return fmt.Errorf("failed to deserialize signature variant: %w", err)
	}

	var expectedLen int
	switch SignatureType(variant) {
	case SignatureTypeSecp256k1:
		expectedLen = SignatureSecp256k1Size
	case SignatureTypeEd25519:
		expectedLen = SignatureEd25519Size
	case SignatureTypeBLS12381:
		expectedLen = SignatureBLS12381Size
	case SignatureTypeFnDsa512:
		expectedLen = SignatureFnDsa512Size
	default:
		return fmt.Errorf("unknown signature variant: %d", variant)
	}

	buf, err := deserializer.DeserializeFixedBytes(expectedLen)
	if err != nil {
		return fmt.Errorf("failed to deserialize signature Bytes: %w", err)
	}

	newSig, err := NewSignatureFromBytes(buf)
	if err != nil {
		return fmt.Errorf("failed to create signature from Bytes: %w", err)
	}

	sig.Variant = SignatureType(variant)
	sig.Bytes = newSig.Bytes
	return nil
}

File: gosdk-develop/python/milon_sdk/crypto/signature.py
"""签名类型与验签（对应 Go crypto/signature.go）。

线上签名长度：Ed25519=64 / Secp256k1=65 / BLS12381=96 / FnDsa512=666。
Postcard 形态：Variant(u32 变长) + 定长字节。
"""
from __future__ import annotations

from enum import IntEnum

import base58  # type: ignore

from ..postcard import Deserializer, Serializer
from . import _bls
from .errors import (
    InvalidPublicKeyError,
    InvalidSignatureError,
    SignatureVerificationError,
)
from .keys import PublicKey, PublicKeyType

SIGNATURE_SECP256K1_SIZE = 65
SIGNATURE_ED25519_SIZE = 64
SIGNATURE_BLS12381_SIZE = 96
SIGNATURE_FNDSA512_SIZE = 666


class SignatureType(IntEnum):
    SECP256K1 = 0
    ED25519 = 1
    BLS12381 = 2
    FNDSA512 = 3


def _signature_size_for(variant: SignatureType) -> int:
    return {
        SignatureType.SECP256K1: SIGNATURE_SECP256K1_SIZE,
        SignatureType.ED25519: SIGNATURE_ED25519_SIZE,
        SignatureType.BLS12381: SIGNATURE_BLS12381_SIZE,
        SignatureType.FNDSA512: SIGNATURE_FNDSA512_SIZE,
    }[SignatureType(variant)]


class Signature:
    __slots__ = ("variant", "bytes")

    def __init__(self, variant: SignatureType, raw: bytes):
        self.variant = SignatureType(variant)
        self.bytes = bytes(raw)

    # ---------------------------------------------------------- 构造
    @classmethod
    def from_bytes(cls, raw: bytes) -> "Signature":
        """按长度自动识别类型。"""
        n = len(raw)
        for st, size in (
            (SignatureType.ED25519, SIGNATURE_ED25519_SIZE),
            (SignatureType.BLS12381, SIGNATURE_BLS12381_SIZE),
            (SignatureType.SECP256K1, SIGNATURE_SECP256K1_SIZE),
            (SignatureType.FNDSA512, SIGNATURE_FNDSA512_SIZE),
        ):
            if n == size:
                return cls(st, raw)
        raise InvalidSignatureError(f"invalid signature length: {n}")

    @classmethod
    def from_string_relaxed(cls, s: str) -> "Signature":
        s = s.strip()
        body = s[2:] if len(s) >= 2 and s[:2] in ("0x", "0X") else s
        try:
            raw = bytes.fromhex(body)
        except ValueError:
            raw = base58.b58decode(s)
            if not raw:
                raise InvalidSignatureError("invalid base58 string")
        return cls.from_bytes(raw)

    # ---------------------------------------------------------- 输出
    def as_bytes(self) -> bytes:
        return self.bytes

    def to_hex(self) -> str:
        return self.bytes.hex()

    def to_base58(self) -> str:
        return base58.b58encode(self.bytes).decode("ascii")

    def __str__(self) -> str:
        return self.to_base58()

    def __repr__(self) -> str:
        return f"Signature({self.variant.name}, {self.to_base58()})"

    def __bytes__(self) -> bytes:
        return self.bytes

    def __eq__(self, other: object) -> bool:
        if isinstance(other, Signature):
            return self.variant == other.variant and self.bytes == other.bytes
        return NotImplemented

    def __hash__(self) -> int:
        return hash((self.variant, self.bytes))

    # ---------------------------------------------------------- 验签
    def verify(self, msg: bytes, pub_key: PublicKey) -> bool:
        if self.variant != pub_key.variant:
            raise InvalidSignatureError(
                f"signature type mismatch: {int(self.variant)} vs {int(pub_key.variant)}"
            )
        if self.variant == SignatureType.SECP256K1:
            return _verify_secp256k1(msg, self.bytes, pub_key)
        if self.variant == SignatureType.ED25519:
            return _verify_ed25519(msg, self.bytes, pub_key)
        if self.variant == SignatureType.BLS12381:
            return _verify_bls12381(msg, self.bytes, pub_key)
        if self.variant == SignatureType.FNDSA512:
            from .fn_dsa512 import verify512

            return verify512(pub_key.as_bytes(), self.bytes, msg)
        return False

    # ---------------------------------------------------------- 编解码
    def marshal_postcard(self, serializer: Serializer) -> None:
        serializer.serialize_u32(int(self.variant))
        serializer.write(self.bytes)

    @classmethod
    def unmarshal_postcard(cls, d: Deserializer) -> "Signature":
        variant = d.deserialize_u32()
        try:
            st = SignatureType(variant)
        except ValueError:
            raise InvalidSignatureError(f"unknown signature variant: {variant}")
        raw = d.deserialize_fixed_bytes(_signature_size_for(st))
        return cls(st, raw)


def _hash_msg_for_secp(msg: bytes) -> bytes:
    """32 字节直通，否则 BLAKE3（与 Go verifySecp256k1 一致）。"""
    if len(msg) == 32:
        return msg
    import blake3  # type: ignore

    return blake3.blake3(msg).digest()  # type: ignore[attr-defined]


def _compact_to_der(sig: bytes) -> bytes:
    """把 64 字节紧凑 R‖S 签名转为 DER（coincurve 的 verify 只接受 DER）。

    对应 Go ethcrypto.VerifySignature 直接吃 R‖S（无 DER 包装）。
    """
    r = int.from_bytes(sig[:32], "big")
    s = int.from_bytes(sig[32:], "big")

    def _int_der(n: int) -> bytes:
        b = n.to_bytes((n.bit_length() + 7) // 8 or 1, "big")
        if b[0] & 0x80:
            b = b"\x00" + b
        return b"\x02" + bytes([len(b)]) + b

    body = _int_der(r) + _int_der(s)
    return b"\x30" + bytes([len(body)]) + body


def _verify_secp256k1(msg: bytes, sig_bytes: bytes, pub_key: PublicKey) -> bool:
    if len(sig_bytes) != SIGNATURE_SECP256K1_SIZE:
        raise InvalidSignatureError(
            f"invalid secp256k1 signature length: expected {SIGNATURE_SECP256K1_SIZE}, "
            f"got {len(sig_bytes)}"
        )
    if not pub_key.is_secp256k1():
        raise InvalidPublicKeyError("not a secp256k1 public key")
    try:
        import coincurve  # type: ignore

        pk = coincurve.PublicKey(pub_key.as_bytes())
        return pk.verify(
            _compact_to_der(sig_bytes[:64]), _hash_msg_for_secp(msg), hasher=None
        )
    except ImportError as exc:  # pragma: no cover
        raise ImportError(f"Secp256k1 验签需要 'coincurve': {exc}") from exc


def _verify_ed25519(msg: bytes, sig_bytes: bytes, pub_key: PublicKey) -> bool:
    if len(sig_bytes) != SIGNATURE_ED25519_SIZE:
        raise InvalidSignatureError(
            f"invalid ed25519 signature length: expected {SIGNATURE_ED25519_SIZE}, "
            f"got {len(sig_bytes)}"
        )
    if not pub_key.is_ed25519():
        raise InvalidPublicKeyError("not an ed25519 public key")
    from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

    try:
        pk = Ed25519PublicKey.from_public_bytes(pub_key.as_bytes())
        pk.verify(sig_bytes, msg)
        return True
    except Exception:
        return False


def _verify_bls12381(msg: bytes, sig_bytes: bytes, pub_key: PublicKey) -> bool:
    if len(sig_bytes) != SIGNATURE_BLS12381_SIZE:
        raise InvalidSignatureError(
            f"invalid BLS signature length: expected {SIGNATURE_BLS12381_SIZE}, "
            f"got {len(sig_bytes)}"
        )
    if not pub_key.is_bls12381():
        raise InvalidPublicKeyError("not a BLS12-381 public key")
    return _bls.bls_verify(pub_key.as_bytes(), sig_bytes, msg)


def verify_batch(
    sigs: list[Signature],
    msgs: list[bytes],
    pub_keys: list[PublicKey],
) -> bool:
    """批量验签：全部为 Ed25519 时逐条验证（Python 无批量加速，语义等价）。"""
    if not (len(sigs) == len(msgs) == len(pub_keys)):
        raise ValueError(
            f"length mismatch: sigs={len(sigs)}, msgs={len(msgs)}, pubkeys={len(pub_keys)}"
        )
    if not sigs:
        return True
    for sig, msg, pk in zip(sigs, msgs, pub_keys):
        if not sig.verify(msg, pk):
            return False
    return True
